### PR TITLE
Add information about default network interface

### PR DESCRIPTION
**Pull request by SeppoTakalo**

### CODE REVIEW

**Comment by SeppoTakalo** [comment]
@AnotherButler Please review

File: docs/reference/api/connectivity/networksocket/networkinterface.md [hunk 1]
```diff
@@ -4,13 +4,73 @@ A socket requires a NetworkInterface instance when opened to indicate which Netw
...
+            "nsapi.default-wifi-ssid": "\"ssid\"",
+            "nsapi.default-wifi-password": "\"password\""
+        }
+    }
+}
+```
+
+Please see [Selecting the default network interface](configuration-connectivity.html#selecting-the-default-network-interface) for information about how to supply required configuration parameters on different connections.
+
+For targets that provide more than one type of connectivity, you may choose the default by overriding the `target.network-default-interface-type` configuration variable.
```
Comment (kjbracey-arm) [inline docs/reference/api/connectivity/networksocket/networkinterface.md]: Insert before this:

> Targets with connectivity set the `target.network-default-interface-type` configuration variable appropriately, either to their only interface or their most-commonly-used one.

Comment (SeppoTakalo) [reply docs/reference/api/connectivity/networksocket/networkinterface.md]: Added.

File: docs/reference/api/connectivity/networksocket/networkinterface.md [hunk 2]
```diff
@@ -4,13 +4,73 @@ A socket requires a NetworkInterface instance when opened to indicate which Netw
...
+
+This example works on all IP-based connectivity options that Mbed OS supports. The Mbed OS configuration system provides configuration for the default interface at build time, so at run time, you can just call `connect()` without any parameters.
+
+For example, when providing Wi-Fi SSID and password, you may use the following `mbed_app.json`:
+
+```
+{
+    "target_overrides": {
+        "*": {
+            "target.network-default-interface-type": "WIFI",
```
[the remark this answers is not in the document]
Comment (SeppoTakalo) [reply docs/reference/api/connectivity/networksocket/networkinterface.md]: Removed

File: docs/reference/api/connectivity/networksocket/networkinterface.md [hunk 3]
```diff
@@ -4,13 +4,73 @@ A socket requires a NetworkInterface instance when opened to indicate which Netw
...
+For targets that provide more than one type of connectivity, you may choose the default by overriding the `target.network-default-interface-type` configuration variable.
+
+Applications may also ask for a specific type of connection, as the following table shows:
+
+|Method|What connectivity is returned| Requirements |
+|------|-----------------------------|--------------|
+|`*WiFiInterface::get_default_instance()`| Wi-Fi interface | Requires security parameters (mode, SSID, password) |
+|`*EthInterface::get_default_instance()` | Wired Ethernet interface, not Wi-Fi | none |
+|`*MeshInterface::get_default_instance()` | Returns either `LoWPANNDInterface` or `ThreadInterface`, depending on which is set to default | Target provides a driver or macro `DEVICE_802_15_4_PHY` is enabled |
+| `*CellularBase::get_default_instance` | Return cellular connectivity | Requires network parameters (pin, APN, username, password) |
```
[the remark this answers is not in the document]
Comment (SeppoTakalo) [reply docs/reference/api/connectivity/networksocket/networkinterface.md]: Fixed

File: docs/reference/api/connectivity/networksocket/networkinterface.md [hunk 4]
```diff
@@ -4,13 +4,73 @@ A socket requires a NetworkInterface instance when opened to indicate which Netw
...
+| `*NetworkInterface::get_default_instance()` | First one from above that is found |  |
+
+Please note that any of those functions may return `NULL` when the interface of this type or its configuration is not found.
```
[the remark this answers is not in the document]
Comment (SeppoTakalo) [reply docs/reference/api/connectivity/networksocket/networkinterface.md]: Thanks, added clarification.

File: docs/reference/api/connectivity/networksocket/networkinterface.md [hunk 5]
```diff
@@ -4,13 +4,73 @@ A socket requires a NetworkInterface instance when opened to indicate which Netw
...
+
+Applications may also ask for a specific type of connection, as the following table shows:
+
+|Method|What connectivity is returned| Requirements |
+|------|-----------------------------|--------------|
+|`*WiFiInterface::get_default_instance()`| Wi-Fi interface | Requires security parameters (mode, SSID, password) |
+|`*EthInterface::get_default_instance()` | Wired Ethernet interface, not Wi-Fi | none |
+|`*MeshInterface::get_default_instance()` | Returns either `LoWPANNDInterface` or `ThreadInterface`, depending on which is set to default | Target provides a driver or macro `DEVICE_802_15_4_PHY` is enabled |
+| `*CellularBase::get_default_instance` | Return cellular connectivity | Requires network parameters (pin, APN, username, password) |
+| `*NetworkInterface::get_default_instance()` | First one from above that is found |  |
```
[the remark this answers is not in the document]
Comment (SeppoTakalo) [reply docs/reference/api/connectivity/networksocket/networkinterface.md]: Modified.

**Comment by SeppoTakalo** [comment]
Clarified the documentation as Kevin suggested.

**Comment by SeppoTakalo** [comment]
@AnotherButler this is now reviewed from engineering point of view. Thanks.

**Comment by SeppoTakalo** [comment]
@AnotherButler Please work with this one. I'm working on delivering changes to external applications regarding default network interface, and I'l like to point users to correct documentation, so I wan't this to be merged.

Also, note that this is OK to be merged into 5.9 branch as well. The API exist there as well. so this can go public.

Thanks.

**Comment by AnotherButler** [comment]
@SeppoTakalo Is there a code dependency on 7814?

**Comment by SeppoTakalo** [comment]
@AnotherButler No code dependency.
7814 just adds more drivers, but the functionality itself is already supported in Mbed OS.

**Pull request merged**